frontend/components/customizer/path-builder.js

```diff
@@ -68,25 +68,41 @@ const NamedParamCaption = styled(BuilderCaption)`
   text-align: center;
 `
 
-export default class PathBuilder extends React.Component {
-  static propTypes = {
-    pattern: PropTypes.string.isRequired,
-    exampleParams: objectOfKeyValuesPropType,
-    onChange: PropTypes.func,
-    isPrefilled: PropTypes.bool,
-  }
-
-  constructor(props) {
-    super(props)
-
-    const { pattern } = props
-    const tokens = pathToRegexp.parse(pattern)
+export function constructPath({ tokens, namedParams }) {
+  let isComplete = true
+  const path = tokens
+    .map(token => {
+      if (typeof token === 'string') {
+        return token
+      } else {
+        const { delimiter, name, optional } = token
+        const value = namedParams[name]
+        if (value) {
+          return `${delimiter}${value}`
+        } else if (optional) {
+          return ''
+        } else {
+          isComplete = false
+          return `${delimiter}:${name}`
+        }
+      }
+    })
+    .join('')
+  return { path, isComplete }
+}
 
-    let namedParams
-    if (this.props.isPrefilled) {
-      namedParams = this.props.exampleParams
+export default function PathBuilder({
+  pattern,
+  exampleParams,
+  onChange,
+  isPrefilled,
+}) {
+  const [tokens] = useState(() => pathToRegexp.parse(pattern))
+  const [namedParams, setNamedParams] = useState(() => {
+    if (isPrefilled) {
+      return exampleParams
```

Comment:
Before, after this general block (if I follow correctly 😄)  `this.state.namedParams` was being set to `exampleParams`. Does that still get achieved with this approach, both if `isPrefilled` is true and if false? Am I ask a crazy/invalid question?

Reply:
As far as I can tell, the old and new code do the same thing, which is to set namedParams to `exampleParams` when `isPrefilled` is true, and otherwise to an object containing an empty string for each named `token`.

The `isPrefilled` case is for badge suggestions, if you want to give it a shot. The logic _looks_ right to me, though I'm not sure I ran it.

Reply:
I wasn't sure if declaring a local `namedParams` on line 105 would impact how everything bubbles up, but will definitely take a look in the UI. Will be much easier for me to poke around the review app

Reply:
Oh yea, I see, it's shadowing the other local. Could rename it to `result` for clarity.

Reply:
I ended up replacing it with a conditional expression and `.reduce()`. Seems to work okay for regular badges and for suggestions.